### ISSUE TITLE
Got Http response code 400 when accessing https://api.sandbox.paypal.com/v1/payments/billing-agreements

### ISSUE THREAD

**Issue by jcardussi** [paypal/PayPal-PHP-SDK#1299]
I got all the way to creating an agreement and got this error:

PayPal\Exception\PayPalConnectionException: Got Http response code 400 when accessing https://api.sandbox.paypal.com/v1/payments/billing-agreements/. in C:\xampp\htdocs\my_php\vendor\paypal\rest-api-sdk-php\lib\PayPal\Core\PayPalHttpConnection.php:207 Stack trace: #0 C:\xampp\htdocs\my_php\vendor\paypal\rest-api-sdk-php\lib\PayPal\Transport\PayPalRestCall.php(78): PayPal\Core\PayPalHttpConnection->execute('{"name":"Base A...') #1 C:\xampp\htdocs\my_php\vendor\paypal\rest-api-sdk-php\lib\PayPal\Common\PayPalResourceModel.php(104): PayPal\Transport\PayPalRestCall->execute(Array, '/v1/payments/bi...', 'POST', '{"name":"Base A...', Array) #2 C:\xampp\htdocs\my_php\vendor\paypal\rest-api-sdk-php\lib\PayPal\Api\Agreement.php(384): PayPal\Common\PayPalResourceModel::executeCall('/v1/payments/bi...', 'POST', '{"name":"Base A...', NULL, Object(PayPal\Rest\ApiContext), Object(PayPal\Transport\PayPalRestCall)) #3 C:\xampp\htdocs\my_php\create-agreement.php(47): PayPal\Api\Agreement->create(Object(PayPal\Rest\ApiContext)) #4 {main}

This seems to be a common problem I have not found a solution that helps me. The plan data seems okay:

{ "id": "P-8A057434G5360803RHRPHZ6A", 
"state": "ACTIVE", 
"name": "Test-Regular1",
 "description": "Create Plan for Regular",
 "type": "INFINITE", 
"create_time": "2019-09-16T23:18:08.376Z", 
"update_time": "2019-09-16T23:18:09.197Z",
"links": [ { "href": "https://api.sandbox.paypal.com/v1/payments/billing-plans/P-8A057434G5360803RHRPHZ6A", "rel": "self", "method": "GET" } ] }

And the code is basically identical to the sample code. I use XXX as a placeholder but I have the real numbers in the actual code:

require "vendor/autoload.php";
use PayPal\Api\Agreement;
use PayPal\Api\Payer;
use PayPal\Api\Plan;
use PayPal\Api\ShippingAddress;

// After Step 1
$apiContext = new \PayPal\Rest\ApiContext(
    new \PayPal\Auth\OAuthTokenCredential(
        'XXX',     // ClientID
        'XXX'      // ClientSecret
    )
);

$agreement = new Agreement();

$agreement->setName('Base Agreement')
    ->setDescription('Basic Agreement')
    ->setStartDate('2019-06-17T9:45:04Z');

$plan = new Plan();
$plan->setId($_GET['planId']);
$agreement->setPlan($plan);

$payer = new Payer();
$payer->setPaymentMethod('paypal');
$agreement->setPayer($payer);

$shippingAddress = new ShippingAddress();
$shippingAddress->setLine1('111 First Street')
    ->setCity('Saratoga')
    ->setState('CA')
    ->setPostalCode('95070')
    ->setCountryCode('US');
$agreement->setShippingAddress($shippingAddress);

try
{
    **$agreement = $agreement->create($apiContext);**
    $token     = $this->getPayPalTokenFromUrl($agreement->getApprovalLink());
} catch (\Exception $e)
{
    echo $e;
}

$approvalUrl = $agreement->getApprovalLink();

header("Location: $approvalUrl");

Any idea?

**Comment by jcardussi** [comment]
I printed more out in hopes that someone can see what is wrong:

$plan = 
{ "id": "P-8A057434G5360803RHRPHZ6A" }

$agreement = 
{ "name": "Base Agreement", "description": "Basic Agreement", "start_date": "2019-06-17T9:45:04Z", "plan": { "id": "P-8A057434G5360803RHRPHZ6A" }, "payer": { "payment_method": "paypal" }, "shipping_address": { "line1": "111 First Street", "city": "Saratoga", "state": "CA", "postal_code": "95070", "country_code": "US" } }

(error code) $e =
PayPal\Exception\PayPalConnectionException: Got Http response code 400 when accessing https://api.sandbox.paypal.com/v1/payments/billing-agreements/. in C:\xampp\htdocs\my_php\vendor\paypal\rest-api-sdk-php\lib\PayPal\Core\PayPalHttpConnection.php:207 Stack trace: #0 C:\xampp\htdocs\my_php\vendor\paypal\rest-api-sdk-php\lib\PayPal\Transport\PayPalRestCall.php(78): PayPal\Core\PayPalHttpConnection->execute('{"name":"Base A...') #1 C:\xampp\htdocs\my_php\vendor\paypal\rest-api-sdk-php\lib\PayPal\Common\PayPalResourceModel.php(104): PayPal\Transport\PayPalRestCall->execute(Array, '/v1/payments/bi...', 'POST', '{"name":"Base A...', Array) #2 C:\xampp\htdocs\my_php\vendor\paypal\rest-api-sdk-php\lib\PayPal\Api\Agreement.php(384): PayPal\Common\PayPalResourceModel::executeCall('/v1/payments/bi...', 'POST', '{"name":"Base A...', NULL, Object(PayPal\Rest\ApiContext), Object(PayPal\Transport\PayPalRestCall)) #3 C:\xampp\htdocs\my_php\create-agreement.php(45): PayPal\Api\Agreement->create(Object(PayPal\Rest\ApiContext)) #4 {main}

**Comment by prakash-gangadharan** [comment]
@jcardussi, what is the merchant email id did you used and help me with the time frame when you executed these call.

**Comment by jcardussi** [comment]
What email are you referring to? My live merchant email ID? Or the sandbox merchant ID? Neither are referred to in any of the code. And the time frame is today. I am rerunning them every single day in hopes that I can get it running. My program is done and waiting to be sold based on being able to sell it.

**Comment by jcardussi** [comment]
Here is the answer: https://github.com/paypal/PayPal-PHP-SDK/wiki/exception-%27PayPal%5CException%5CPayPalConnectionException%27-with-message-%27Got-Http-response-code-400-when-accessing

It prints out the actual error. Why is this not in the sample code? It would have saved me days of digging.

**Comment by jcardussi** [comment]
For the record: My problem was with the start up date (has to be later than now).

**Comment by prakash-gangadharan** [comment]
@jcardussi, so you got the `paypal-debud-id` now and can you please help me with that? and I asked the sandbox merchant ID for which you have created the client-id and secret.

**Comment by jcardussi** [comment]
If I post it here, it will be available to the world and it includes my paypal sign in ID since I cannot change it.

**Comment by jcardussi** [comment]
This code is so much better than what is in the sample code:

try {
    $payment->create($apiContext);
} catch (PayPal\Exception\PayPalConnectionException $ex) {
    echo $ex->getCode(); // Prints the Error Code
    echo $ex->getData(); // Prints the detailed error message 
    die($ex);
} catch (Exception $ex) {
    die($ex);
}

I dropped it in and it gave a much better error message.

**Comment by prakash-gangadharan** [comment]
@jcardussi, did you manage to find what is the flaw with your request? or still you facing any challenges?

**Comment by jcardussi** [comment]
The flaw was my start date was not in the future. This bug is solved.

**Comment by usamamashkoor** [comment]
> For the record: My problem was with the start up date (has to be later than now).

Hi, I am facing the same issue how can I set up my start date so that I can solve this issue?

**Comment by jcardussi** [comment]
I am not sure but I think I set it up for 30 days out and an immediate payment. Let me know if you are stuck.

**Comment by usamamashkoor** [comment]
@jcardussi Thanks I appreciate your help, I have solved the issue in case it was a description which was causing the issue.

Thanks again :)
